onnxruntime/core/framework/stream_execution_context.cc

```diff
@@ -0,0 +1,235 @@
+                                                                           count_down_barriers_(num_barriers),
+                                                                           single_thread_mode_(single_thread_mode) {
+  notifications_.reserve(notification_owners.size());
+  for (size_t i = 0; i < notification_owners.size(); ++i) {
+    auto* stream = device_stream_map_.GetStream(notification_owners[i]);
+    if (stream)
+      notifications_.emplace_back(stream->CreateNotification(/*TODO: calculate num of consumers*/ 0));
+    else
+      notifications_.push_back(nullptr);
+  }
+  std::atomic_int* p_release_plan_buffer = new std::atomic_int[sess_state.GetExecutionPlan()->release_actions.size()];
+  release_plan_ = std::unique_ptr<std::atomic_int[]>(p_release_plan_buffer);
+
+  // init barreris
+  for (size_t i = 0; i < num_barriers; ++i) {
+    count_down_barriers_[i].Set(2);
+  }
+  // init remain task to number of streams
+  remain_tasks_.Set(num_streams);
+  // generate release plan (the ref counts)
+  auto& release_actions = sess_state.GetExecutionPlan()->release_actions;
+  for (size_t i = 0; i < release_actions.size(); ++i) {
+    release_plan_[i] = static_cast<int>(release_actions[i].ref_count);
+  }
+}
+
+const SessionState& StreamExecutionContext ::GetSessionState() const { return *session_state_; }
+
+const logging::Logger& StreamExecutionContext ::GetLogger() const { return *logger_; }
+
+ExecutionFrame& StreamExecutionContext ::GetExecutionFrame() { return frame_; }
+
+synchronize::Notification* StreamExecutionContext ::GetNotification(size_t idx) { return notifications_[idx].get(); }
+
+bool StreamExecutionContext ::DecCountDownBarrier(size_t barrier_id) {
+  return count_down_barriers_[barrier_id].Dec();
+}
+
+Stream* StreamExecutionContext ::GetDeviceStream(size_t idx) {
+  ORT_ENFORCE(idx < device_stream_map_.NumStreams());
+  return device_stream_map_.GetStream(idx);
+}
+
+const Status& StreamExecutionContext ::TaskStatus() const {
+  return task_status_;
+}
+
+void StreamExecutionContext ::CompleteTask() {
+  remain_tasks_.Dec();
+}
+
+void StreamExecutionContext ::AddTask() {
+  remain_tasks_.Inc();
+}
+
+void StreamExecutionContext ::WaitAll() {
+  while (remain_tasks_.Get())
+    onnxruntime::concurrency::SpinPause();
+}
+
+void StreamExecutionContext ::SetStatus(Status& status) {
+  // TODO: if multiple worker report non-ok status,
+  // what is our strategy? currently we just keep
+  // a random one. as long as it is not OK, the
+  // execution will fail.
+  if (task_status_.IsOK() && !status.IsOK())
+    task_status_ = status;
+}
+
+StreamExecutionContext ::~StreamExecutionContext() {}
+
+void StreamExecutionContext ::RecycleNodeInputs(onnxruntime::NodeIndex node_index) {
+  auto* execution_plan = session_state_->GetExecutionPlan();
+  for (auto idx : execution_plan->node_release_list[node_index]) {
+    if (--release_plan_[idx] == 0) {
```

Comment:
My understanding is release_plan_[idx] equals execution_plan->release_actions[idx].ref_count, true? Then why not use ref_count directly? (I don't see any place to that ref_count needs to be reserved)

Reply:
ref_count in execution plan is the immutable status that shared across multiple parallel execution, you can't & shouldn't modify it in an execution thread.